src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1167,9 +1177,45 @@ static RegexNode ExtractCommonPrefixOneNotoneSet(RegexNode alternation)
                     children.RemoveRange(startingIndex + 1, endingIndex - startingIndex - 1);
                 }
 
-                // If we've reduced this alternation to just a single branch, return it.
-                // Otherwise, return the alternation.
-                return alternation.ChildCount() == 1 ? alternation.Child(0) : alternation;
+                return alternation.ReplaceNodeIfUnnecessary();
+            }
+
+            // Removes unnecessary Empty and Nothing nodes from the alternation. A Nothing will never
+            // match, so it can be removed entirely, and an Empty can be removed if there's a previous
+            // Empty in the alternation: it's an extreme case of just having a repeated branch in an
+            // alternation, and while we don't check for all duplicates, checking for empty is easy.
+            static RegexNode RemoveRedundantEmptiesAndNothings(RegexNode node)
+            {
+                Debug.Assert(node.Type == Alternate);
+                Debug.Assert(node.ChildCount() >= 2);
+                var children = (List<RegexNode>)node.Children!;
+
+                int i = 0, j = 0;
+                bool seenEmpty = false;
```

Comment:
Shouldn't we also have a seenNothing that is handled specially? TBH I'm not sure what is the difference between a Nothing and an Empty in an alternation, won't both be expressed as empty strings?

Reply:
Effectively both Empty and Nothing match an empty string but differ what they do for one: Empty matches, Nothing fails. So if you have a branch in an alternation that's entirely a Nothing, it won't ever succeed, so we can just remove the whole branch. In contrast, an Empty branch will always succeed, so we can't remove it... but if there's a duplicate branch later, we can remove the duplicate.

Reply:
so in that case shouldn't we need to have a seenNothing bool and have a specific case for `case Nothing when !seenNothing` bellow similar to what we do with Empty? Or is an Alternation Node only ever going to have at most one `Nothing` as a child?

Reply:
We can remove every Nothing.  We need to keep the first Empty.  That's why we have `seenEmpty`.

For example, if you have `(ab||cd||ef)gh`, which has an `ab` branch, an empty branch, a `cd` branch, another empty branch, and an `ef` branch, it's valid to transform that into `(ab||cd|ef)gh`, but it's not valid to transform that into `(ab|cd|ef)gh`.  The former will match "gh" (following the empty branch) but the latter won't (because all the empties were erroneously removed).

In contrast, if you have `(ab|(?!)|cd|(?!)|ef)gh`, where we have nothing branches instead of the empties, it _is_ valid to transform that into `(ab|cd|ef)gh`, because nothing will ever match any of the nothing branches.